packages/SwingSet/src/liveslots/liveslots.js

```diff
@@ -1496,6 +1496,7 @@ function build(
     vom.flushCache();
     await gcTools.gcAndFinalize();
     const doMore = await scanForDeadObjects();
+    // @ts-expect-error FIXME doMore is void
```

Comment:
Hm, this looks like both functions think they're responsible for the re-scan.. I bet we didn't think this through completely when we introduced `scanForDeadObjects()`.

The `doMore` test is meant to deal with the case where `gcAndFinalize` shakes loose some in-RAM `Representative`s,  `scanForDeadObjects` looks through `possiblyDeadSet` and finds some actual vrefs it can retire (`vrm.possibleVirtualObjectDeath()`), some of those vrefs were used as keys in a (virtual) WeakMap, so deleting the keys can drop references to more in-RAM Representatives. We need a new `gcAndFinalize` pass to shake those loose, so we repeat the loop within `scanForDeadObjects()` until `possibleVirtualObjectDeath()` didn't free anything (and our other checks didn't free anything either).

The outer function(`bringOutYourDead`) looks like it was meant to do the same "repeat until nothing changes" pattern, although with recursion rather than a `do/while` loop. It does a (redundant?) extra `gcAndFinalize` at the expected time. I think the only real contribution of the outer loop, these days, is an extra call to `vom.flushCache()`, but I'm pretty sure that can't shake loose more objects like `doMore` is tracking, so I think it's not important to call from there.

@FUDCo is the authority here, but I'm thinking there are two correct fixes:

* 1: remove the `gcAndFinalize` and the recursion from `bringOutYourDead` (makie it just do `vom.flushCache(); await gcTools.gcAndFinalize(); await scanForDeadObjects(); return`, and rely upon the loop that happens inside `scanForDeadObjects`
* 2: remove the `do/while` from `scanForDeadObjects`, and modify it to return `doMore`, relying upon the recursion in `bringOutYourDead` to shake everything loose
  * if we go this way, I think I'd prefer to have `bringOutYourDead` use a do/while loop, instead of the recursion, since we can't rely on tail-call optimization

The benefit of doing the loop inside `scanForDeadObjects` (1) is that we only emit a single trio of syscalls, no matter how many times we have to loop. I can't think of a benefit for (2), other than that (1) makes `bringOutYourDead` start to look kind of useless and suggests we might refactor something

Reply:
To unblock the other `checkJs` PRs I propose that this land with the FIXME comment and the package maintainers prioritize fixing it as they wish.